File: twoIntegerSumII.cpp
//two pointer
//time complexity -> O(n) uses while loop
//space complexity -> O(1) we used two pointer and doesn't use any extra memory
class Solution{
public:
    vector<int> twoSum(vector<int>& numbers, int target){
        int l = 0;
        int r = numbers.size()-1;
        while (l < r){
            if(numbers[l] + numbers[r] == target){
                return {l + 1, r + 1};
            }
            else if (numbers[l] + numbers[r] > target){
                r--;
            }
            else{ // two values from pointers are less than target
                l++;
            }
        }
    }
};

///////////////////////////////////////////////
///////////////////////////////////////////////
//hashmap
//Time complexity -> O(n) due to using for loop
//space complexity -> O(n) in memory because unordered map uses memory, uses data structure
#include <unordered_map>
class Solution{
    vector<int> twoSum(vector<int>& numbers, int target){
        std::unordered_map<int, int> hashmap;
        for(int i = 0; i < numbers.size(); i++){
            temp = target - numbers[i];
            if(hashmap.count(temp)){
                return {hashmap[temp] + 1, numbers[i] + 1}
            }
            hashmap[numbers[i]] = i;
        }
    }
}

File: twoIntegerSumII.py
#two pointers
#time complexity -> O(n)
#space complexity -> O(1)
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        l = 0
        r = len(numbers)-1

        while(l < r):
            if numbers[l] + numbers[r] == target:
                return [l+1, r+1]
            elif numbers[l] + numbers[r] > target:
                r -= 1
            else: #value of two pointers is less than target
                l += 1

        return []


#####################################
#####################################
#hashmap
#time compelxity -> O(n)
#space complextiy -> O(n) due to using defaultdict (hashmap) in memory
class Solution:
    def twoSum(self, numbers: List[int], target: int):
        hashmap = defaultdict(int) #dictionary
        for i in range(len(numbers)):
            temp = target - i
            if hashmap[temp]:
                return [hashmap[temp], i + 1]
            hashmap[numbers[i]] = i + 1
        return [] 